launcher/src/main/java/org/apache/spark/launcher/LauncherConnection.java

```diff
@@ -95,15 +95,15 @@ protected synchronized void send(Message msg) throws IOException {
   }
 
   @Override
-  public void close() throws IOException {
+  public synchronized void close() throws IOException {
     if (!closed) {
-      synchronized (this) {
```

Comment:
what's wrong with this? It's a classic "double-checked locking" in java.

Reply:
The sub-class needs coarser-grained locking so this wasn't really doing anything useful.